common/autoware_motion_utils/include/autoware/motion_utils/trajectory_container/interpolator/interpolator.hpp

```diff
@@ -0,0 +1,261 @@
+   *
+   * @param axis The axis values.
+   * @param values The values to interpolate.
+   * @throw std::runtime_error if the inputs are invalid.
+   */
+  void validate_build_input(
+    const Eigen::Ref<const Eigen::VectorXd> & axis, const std::vector<T> & values) const
+  {
+    if (static_cast<size_t>(axis.size()) != values.size()) {
+      throw std::runtime_error(
+        fmt::format(
+          "Axis and values must have the same size. Axis size: {}, values size: {}.", axis.size(),
+          values.size())
+          .c_str());
+    }
+    if (static_cast<size_t>(axis.size()) < minimum_required_points()) {
+      throw std::runtime_error(
+        fmt::format(
+          "Axis and values must have at least {} points. Axis size: {}, values size: {}.",
+          minimum_required_points(), axis.size(), values.size())
+          .c_str());
+    }
+  }
+
+  /**
+   * @brief Validate the input to the compute method.
+   *
+   * Checks that the interpolator has been built and that the input value is within range.
+   *
+   * @param s The input value.
+   * @throw std::runtime_error if the interpolator has not been built.
+   */
+  void validate_compute_input(const double & s) const
+  {
+    if (!is_built_) {
+      throw std::runtime_error("Interpolator has not been built yet.");
+    }
+    if (s < start() || s > end()) {
+      RCLCPP_WARN(
+        rclcpp::get_logger("InterpolatorBase"),
+        "Input value %f is outside the range of the interpolator [%f, %f].", s, start(), end());
+    }
+  }
+
+public:
+  /**
+   * @brief Build the interpolator with the given axis and values.
+   *
+   * @param axis The axis values.
+   * @param values The values to interpolate.
+   * @return Reference to the interpolator.
+   */
+  InterpolatorCommonImpl & build(
+    const Eigen::Ref<const Eigen::VectorXd> & axis, const std::vector<T> & values)
+  {
+    validate_build_input(axis, values);
+    this->axis_ = axis;
+    build_impl(values);
+    this->is_built_ = true;
+    return *this;
+  };
+
+  /**
+   * @brief Build the interpolator with the given axis and values.
+   *
+   * @param axis The axis values.
+   * @param values The values to interpolate.
+   * @return Reference to the interpolator.
+   */
+  InterpolatorCommonImpl & build(const std::vector<double> & axis, const std::vector<T> & values)
+  {
+    Eigen::Map<const Eigen::VectorXd> axis_map(axis.data(), static_cast<Eigen::Index>(axis.size()));
+    return build(axis_map, values);
+  }
+
+  /**
+   * @brief Get the minimum number of required points for the interpolator.
+   *
+   * This method should be overridden by subclasses to return the specific requirement.
+   *
+   * @return The minimum number of required points.
+   */
+  [[nodiscard]] virtual size_t minimum_required_points() const = 0;
+
+  [[nodiscard]] int get_index(const double & s, bool include_end = true) const
+  {
+    if (include_end && s == end()) {
+      return static_cast<int>(axis_.size()) - 2;
```

Comment:
I'd like to provide some comments for improvement. Although `int` is frequently used in Autoware, it is actualy an unsafe type. Please use `uint32_t` instaed.
cf. https://matt.sh/howto-c#:~:text=Writing%20code-,Types,-If%20you%20find

Reply:
Thank you for your review!! I will fix it.

Reply:
I have fixed it. This index is used by eigen, so I use `int32_t`.